File: 2015/day5.ts
import { readFileSync } from "fs";

const data = readFileSync("./inputs/day5.txt", "utf-8");

const p1rule1Regex = /[aeiou]/g;
function p1rule1(val: string): boolean {
  const result = val.match(p1rule1Regex);
  return !!result && result.length > 2;
}

function p1rule2(val: string): boolean {
  return !!val.split("").find((v, i, arr) => v == arr[i + 1]);
}

const p1rule3Regex = /ab|cd|pq|xy/;
function p1rule3(val: string): boolean {
  return !p1rule3Regex.test(val);
}

function part1() {
  const nice = data
    .split("\n")
    .filter((v) => p1rule1(v) && p1rule2(v) && p1rule3(v));
  console.log(`P1 Nice: ${nice.length}`);
}

function p2rule1(val: string): boolean {
  return !!val
    .split("")
    .find((_, i) => val.indexOf(val.slice(i, i + 2), i + 2) > 0);
}

function p2rule2(val: string): boolean {
  return !!val.split("").find((v, i, arr) => v == arr[i + 2]);
}

function part2() {
  const nice = data.split("\n").filter((v) => p2rule1(v) && p2rule2(v));
  console.log(`P1 Nice: ${nice.length}`);
}

console.log("DAY 5");
part1();
part2();

File: 2020/day5.py
data = ""
with open("inputs/day5.txt", "r") as f:
    data = f.read()


data = set(
    int("".join(["0" if c == "F" or c == "L" else "1" for c in s]), 2)
    for s in data.split("\n")
)


def Part1() -> None:
    print("Max Seat: {}".format(max(data)))


def Part2() -> None:
    for i in range(min(data) + 1, max(data)):
        if i not in data:
            print("My Seat: {}".format(i))
            break


def main() -> None:
    print("Day 5")
    Part1()
    Part2()


if __name__ == "__main__":
    main()
